### PR TITLE
 pkgs/top-level: init expression for haskell-updates jobset 

### PR DESCRIPTION
**Pull request by sternenseemann**
release-haskell.nix is intended to be a replacement for
https://github.com/peti/ci/blob/master/haskell-nixpkgs.nix
which is currently the main expression for the haskell-updates jobset
on hydra (in the nixpkgs project).

It has the same jobs as the old haskell-nixpkgs.nix file:

* haskellPackages.*
* haskell.compiler.*
* Some extra haskell packages for certain compilers

The following jobs are new:

* tests.haskell.*
* A manually maintained list of top-level haskell packages (most of them
  using justStaticExecutables)
* An aggregate job which is intended to aid merging the haskell-updates
  branch: It holds an arbitrary list of haskell-related packages and
  tests we intend have working at all times. This is still somewhat
  incomplete and should be extendend in the future.

Additionally a lot of refactoring has been done and some unnecessary
code has been eliminated. Due to the increased set of jobs and my
ideas of convenience however, the code size has grown overall.
I've tried document the individual parts and would be happy about
feedback in general.

One future improvement could be making adding top-level haskell packages
more convenient and adding them all to the aggregate job automatically.

Evaluation is currently at 20s on my laptop, I suspect the main offender is
`versionedCompilerJobs`, but haven't done any testing

---

Contains some commits of #121386.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by maralorn** [comment]
This looks very good to me.

I am not enough of an expert to judge if this actually does exactly what we want on hydra. But I’ll get we will figure that out.

I like the mergeable job a lot. My suggestion would be to add the top-level package haskell-language-server in there.

File: pkgs/top-level/release-haskell.nix
Comment (maralorn) [inline pkgs/top-level/release-haskell.nix:159]: What do you think about using the
```
builtins.attrValues {
 inherit (jobs) cabal-install ...;
 inherit (jobs.haskellPackages) cabal-plan ...;
}
```
idiom?

**Review by maralorn — APPROVED** [review]
I think this is nice. Let’s try it and improve on it, if we notice any problems.

File: pkgs/top-level/release-haskell.nix
Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix:159]: This doesn't work as nicely because `constituents` doesn't like attribute sets of derivations. However every primitive job is actually a set of the form { x86_64-linux = drv; }. I've written `accumulateDerivations` in an effort of preventing us from specifying all platforms manually in the future if we add another platform to `supportedSystems`.

OTOH we could just not care about this and explicitly state the platforms as is done in `pkgs/top-level/release.nix`, for example.

Comment (maralorn) [reply pkgs/top-level/release-haskell.nix:165]: I didn‘t mean anything so complicated.

Can‘t we just write
`constituents = accumulateDerivations (builtins.attrValue { ... })`

That way we can avoid repeating "jobs" and "jobs.haskellPackages".

Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix:165]: I just wanted to keep it in the spirit of the original API of `releaseTools.aggregate`.

Comment (maralorn) [reply pkgs/top-level/release-haskell.nix:165]: I feel like we are talking past each other. But it’s not important. You can just leave it, as it is.

Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix:165]: I guess we can talk tomorrow just as well :)

File: pkgs/top-level/release-haskell.nix
Comment (cdepillabout) [inline pkgs/top-level/release-haskell.nix:161]: Should all three of us be here (as well as below)?

At the very least, I'd be happy to be here.

File: pkgs/top-level/release-haskell.nix
Comment (cdepillabout) [inline pkgs/top-level/release-haskell.nix:3]: Would you be able to expand on this slightly?  For instance, what package is `hydra-eval-jobs` provided by?

On my NixOS system, `command-not-found` tells me that `hydra-eval-jobs` is provided by both the `hydra-migration` and `hydra-unstable` derivations.

Also, does this `hydra-eval-jobs` command kick off an actual job on Hydra?  Or does this just compile everything locally?

**Comment by cdepillabout** [comment]
This looks good to me, although I really don't have enough experience with Hydra to give an in-depth review.

File: pkgs/top-level/release-haskell.nix
Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix:3]: Sure, I can another comment. You need `hydra-eval-jobs` from `hydra-unstable` and all the command does is print a jobset description in JSON format to `stdout`.

File: pkgs/top-level/release-haskell.nix
Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix]: I just went ahead and added a haskell maintainer team in `lib.teams`.

File: pkgs/top-level/release-haskell.nix
Comment (cdepillabout) [inline pkgs/top-level/release-haskell.nix]: Ah, good idea!

File: pkgs/top-level/release-haskell.nix
Comment (andir) [inline pkgs/top-level/release-haskell.nix:8]: Shouldn't this be the same as with the other release jobs? I am thinking that at least `aarch64-linux` should be in here as well.

File: pkgs/top-level/release-haskell.nix
Comment (andir) [inline pkgs/top-level/release-haskell.nix:148]: What is the requirement for a job to be in this section? What does it mean when one of them breaks on on compiler but not on another?

File: pkgs/top-level/release-haskell.nix
Comment (andir) [inline pkgs/top-level/release-haskell.nix:160]: Why is this called `mergeable`? I guess this is the equivalent of the `tested` jobset for haskell? The description could probably be better here.

File: pkgs/top-level/release-haskell.nix
Comment (andir) [inline pkgs/top-level/release-haskell.nix:196]: What are the consequences of a build failure for packages in this category? Does that mean someone has to look? Mind specifying that here (in he description).

**Review by andir — COMMENTED** [review]
Doesn't look too bad. Just some quality of life comments. I started a test eval on my private hydra instance: https://hydra.h4ck.space/jobset/nixpkgs/haskell-update#tabs-evaluations (ipv6-only)

File: pkgs/top-level/release-haskell.nix
Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix:8]: This is like this for historical reasons. Since haskell packages get built on `aarch64-linux` and `x86_64-darwin` in the main jobsets as well, I think we can enable it here as well. I just want to poke the infra team if it's okay.

File: pkgs/top-level/release-haskell.nix
Comment (cdepillabout) [reply pkgs/top-level/release-haskell.nix:196]: We talked about this shortly in our meeting today:

This is an aggregate job that contains all Haskell packages that have maintainer.  If one of these packages is not building, then we should mark the package as `broken` and leave the maintainer as-is.  We also need to ping them on the PR for the `haskell-updates` branch so that they have time to fix the problem.  

We agreed that we need to give the maintainers at least 7 days to fix their packages before we mark them broken and merge `haskell-updates` into `master`.

File: pkgs/top-level/release-haskell.nix
Comment (cdepillabout) [reply pkgs/top-level/release-haskell.nix:160]: We talked about this shortly in our meeting:

This is an aggregate job that contains all "important" Haskell packages.  If this is not passing, then we should _not_ merge haskell-updates into `master`.  In Hydra UI this shows up as a single job.

**Comment by maralorn** [comment]
So, @sternenseemann do you wanna add a bit more comments regarding to those questions? Or do we just merge (with the intention to document the meaning of those jobs somewhere else)?

**Comment by cdepillabout** [comment]
Ah, one other request:  could you add the file in this PR to both `.github/CODEOWNERS` and `.github/labeler.yml` so that we get notified when it is changed?

File: pkgs/top-level/release-haskell.nix
Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix:148]: I've elaborated on this a bit in the commit. I think what warrants inclusion will become clearer in the future, as we may want to expand the list of packages there.

File: pkgs/top-level/release-haskell.nix
Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix:166]: I've elaborated the description, I hope Hydra can deal with newlines in that field.

File: pkgs/top-level/release-haskell.nix
Comment (sternenseemann) [reply pkgs/top-level/release-haskell.nix:205]: I wouldn't prescribe a consequence of this at this time, for now I think this is mostly a nice source for an overview.

**Comment by maralorn** [comment]
Okay, let’s do this!

**Comment by maralorn** [comment]
Thx!

**Pull request merged**
